WTG.Analyzers/Analyzers/PublicTupleType/PublicTupleTypeAnalyzer.cs

```diff
@@ -0,0 +1,303 @@
+using System;
+using System.Collections.Immutable;
+using System.Threading;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using WTG.Analyzers.Utils;
+
+namespace WTG.Analyzers
+{
+	[DiagnosticAnalyzer(LanguageNames.CSharp)]
+	public sealed partial class PublicTupleTypeAnalyzer : DiagnosticAnalyzer
+	{
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(
+			Rules.AvoidTupleTypesInPublicInterfacesRule);
+
+		public override void Initialize(AnalysisContext context)
+		{
+			context.EnableConcurrentExecution();
+			context.RegisterCompilationStartAction(CompilationStart);
+		}
+
+		void CompilationStart(CompilationStartAnalysisContext context)
+		{
+			var cache = new FileDetailCache();
+
+			context.RegisterSyntaxNodeAction(
+				c => Analyze(c, cache),
+				SyntaxKind.ConstructorDeclaration,
+				SyntaxKind.MethodDeclaration,
+				SyntaxKind.PropertyDeclaration,
+				SyntaxKind.IndexerDeclaration,
+				SyntaxKind.DelegateDeclaration,
+				SyntaxKind.EventDeclaration,
+				SyntaxKind.EventFieldDeclaration,
+				SyntaxKind.FieldDeclaration);
+		}
+
+		static void Analyze(SyntaxNodeAnalysisContext context, FileDetailCache cache)
+		{
+			if (cache.IsGenerated(context.SemanticModel.SyntaxTree, context.CancellationToken))
+			{
+				return;
+			}
+
+			var decl = (MemberDeclarationSyntax)context.Node;
+			var processor = new Processor(context.SemanticModel, context.ReportDiagnostic, context.CancellationToken);
+			decl.Accept(processor);
+		}
+
+		sealed class Processor : CSharpSyntaxVisitor
+		{
+			public Processor(SemanticModel model, Action<Diagnostic> report, CancellationToken cancellationToken)
+			{
+				this.model = model;
+				this.report = report;
+				this.cancellationToken = cancellationToken;
+			}
+
+			public override void VisitConstructorDeclaration(ConstructorDeclarationSyntax node)
+			{
+				var syntax = model.GetDeclaredSymbol(node, cancellationToken);
+
+				if (syntax != null && syntax.IsExternallyVisible())
+				{
+					VisitParameters(node.ParameterList.Parameters);
+				}
+			}
+
+			public override void VisitMethodDeclaration(MethodDeclarationSyntax node)
+			{
+				if (!IsOverride(node.Modifiers))
```

Comment:
I usually prefer code to follow the happy path by returning early on preconditions instead of indenting everything extra levels.

Reply:
I tend to think of it as "in this particular situation, we do X, in all other situations we do Y", and then avoid repeating 'Y' (even though in this case, 'Y' is empty.)